### PR TITLE
Make Discord notifications pretty

### PR DESCRIPTION
**Pull request by ACA30**
## Describe your changes

Make Discord webhook notifications prettier by using embeds! Supports the same content as before, but adds support for images for Discord notifications.

image
image

NOTE: A Twitch emote should be added to the Mayacord server (and the ID updated in this PR) - this is from one of my personal testing servers, and I'm not sure if Discord will reliably let the webhook use the emote - better to be safe and have the emote in the same server as the webhook.

## Notes for testing your change

Images will be broken on localhost (since Discord can't load your images) - test from staging or manually edit code to get images from somewhere else.

### CODE REVIEW

File: apps/website/src/server/discord.ts [hunk 1]
```diff
@@ -1,52 +1,93 @@
...
+    color: 0x636a60, // Custom color
+    fields: contentLink
+      ? [
+          {
+            name: "Link",
+            value: formattedLink,
+            inline: true,
+          },
+        ]
+      : [],
```
Comment (MattIPv4) [inline apps/website/src/server/discord.ts:64]:
```suggestion
    url: contentLink,
    description: contentMessage || "",
    color: 0x636a60, // Custom color
```
Related to the comment above, I think a dedicated link field is also a bit excessive, I think we can just include the link as part of the embed directly, which will make the title the link?

File: apps/website/src/server/discord.ts [hunk 2]
```diff
@@ -1,52 +1,93 @@
...
   botName?: string;
   expiresAt?: Date;
   toEveryone?: boolean;
 }) {
+  const linkRegex = /twitch\.tv\/(\w+)/;
+  const match = contentLink ? contentLink.match(linkRegex) : null;
+  const formattedLink = match
+    ? `[<:twitch:603947671941546025>/${match[1]}](${contentLink})`
+    : `[Click Here](${contentLink})`;
+
```
Comment (MattIPv4) [inline apps/website/src/server/discord.ts:51]:
```suggestion
```
I think having the Twitch emote here is probably a bit excessive, and using `click here` is generally bad practice as it is not descriptive.

File: apps/website/src/server/discord.ts [hunk 3]
```diff
@@ -1,52 +1,93 @@
...
+      ? [
+          {
+            name: "Link",
+            value: formattedLink,
+            inline: true,
+          },
+        ]
+      : [],
+    footer: {
+      text: "Alveus Sanctuary • Notifications",
```
Comment (MattIPv4) [inline apps/website/src/server/discord.ts:66]:
```suggestion
      text: "Alveus Sanctuary • Notifications • alveus.gg/updates",
```

File: apps/website/src/server/notifications.ts [hunk 4]
```diff
@@ -209,15 +210,12 @@ async function createDiscordNotifications({
 
   const tasks = [];
   for (const webhookUrl of webhookUrls) {
-    const relativeNotificationUrl = `/notifications/${id}`;
-    const link =
-      linkUrl ||
-      new URL(relativeNotificationUrl, env.NEXT_PUBLIC_BASE_URL).toString();
```
Comment (MattIPv4) [inline apps/website/src/server/notifications.ts:215]: We seem to have lost this logic here for defaulting to the notification URL if one isn't set?

File: apps/website/src/server/notifications.ts [hunk 5]
```diff
@@ -209,15 +210,12 @@ async function createDiscordNotifications({
...
-    const content = `${title}\n${message}\n${link}`;
-
     tasks.push(
       triggerDiscordChannelWebhook({
-        content,
+        contentTitle: title || undefined,
```
Comment (MattIPv4) [inline apps/website/src/server/notifications.ts:215]:
```suggestion
        contentTitle: title ?? undefined,
```
So we don't cast a falsey string to undefined

File: apps/website/src/server/notifications.ts [hunk 6]
```diff
@@ -209,15 +210,12 @@ async function createDiscordNotifications({
...
+        contentMessage: message || undefined,
+        contentLink: linkUrl || undefined,
+        imageUrl: imageUrl || undefined,
```
Comment (MattIPv4) [inline apps/website/src/server/notifications.ts:218]:
```suggestion
        contentLink: linkUrl ?? undefined,
        imageUrl: imageUrl ?? undefined,
```
Ditto

File: apps/website/src/server/notifications.ts [hunk 7]
```diff
@@ -209,15 +210,12 @@ async function createDiscordNotifications({
...
-    const link =
-      linkUrl ||
-      new URL(relativeNotificationUrl, env.NEXT_PUBLIC_BASE_URL).toString();
-    const content = `${title}\n${message}\n${link}`;
-
     tasks.push(
       triggerDiscordChannelWebhook({
-        content,
+        contentTitle: title || undefined,
+        contentMessage: message || undefined,
```
Comment (MattIPv4) [inline apps/website/src/server/notifications.ts:216]:
```suggestion
        contentMessage: message,
```
message is required in the notification schema

**Review by MattIPv4 — COMMENTED** [review]
@pjeweb overall thoughts on whether this is a beneficial change? I'm not sure how embeds render in Discord push notifs?

**Comment by ACA30** [comment]
> I'm not sure how embeds render in Discord push notifs?

They do still render only the content
image

Perhaps instead of the link remaining in the base content field it should be the title? That keeps notifications more clear but still allows the embed as the main content.

---

One of the reasons I like embeds is because sometimes Discord just renders links weirdly (for example, arbitrarily deciding to sometimes use thumbnail and sometimes use image??)
image

Maybe stylizing more like the Streamcord notifications could look nice (and be more consistent)? That button would be nice to add.
image

**Comment by ACA30** [comment]
I could add the "Watch Stream" button as well, as seen in the Streamcord notifications, but that requires an application-owned webhook ([docs](https://discord.com/developers/docs/resources/webhook#execute-webhook-jsonform-params)).

File: apps/website/src/server/discord.ts [hunk 8]
```diff
@@ -1,52 +1,81 @@
...
+  contentMessage?: string;
+  contentLink?: string;
+  imageUrl?: string;
   botName?: string;
   expiresAt?: Date;
   toEveryone?: boolean;
 }) {
+  const embed = {
+    title: contentTitle || "Notification",
+    description: contentMessage || "",
```
Comment (MattIPv4) [inline apps/website/src/server/discord.ts:49]:
```suggestion
    description: ((contentMessage || "") + (contentLink ? `\n[${contentLink.replace(/^https?:\/\//(www\.)?, "")}](${contentLink})` : "")).trim(),
```
Thoughts on having the link in the content as well for clarity?

**Review by MattIPv4 — DISMISSED** [review]
LGTM. @pjeweb any thoughts?

**Comment by pjeweb** [comment]
Looks great in terms of screenshots and code. Have not tested it yet and I do not think i can properly test this before next week. Im still on workation.

 @MattIPv4  if you deem it working, feel free to merge

File: apps/website/src/server/discord.ts [hunk 9]
```diff
@@ -1,52 +1,81 @@
...
+  contentMessage?: string;
+  contentLink?: string;
+  imageUrl?: string;
   botName?: string;
   expiresAt?: Date;
   toEveryone?: boolean;
 }) {
+  const embed = {
+    title: contentTitle || "Notification",
+    description: ((contentMessage || "") + (contentLink ? `\n[${contentLink.replace(/^https?:\/\//(www\.)?, "")}](${contentLink})` : "")).trim(),
```
Comment (MattIPv4) [inline apps/website/src/server/discord.ts:49]:
```suggestion
    description: ((contentMessage || "") + (contentLink ? `\n[${contentLink.replace(/^https?:\/\/(www\.)?/, "")}](${contentLink})` : "")).trim(),
```
My bad (plus whatever prettier needs so you might want to do this locally)

**Review by MattIPv4 — APPROVED** [review]
LGTM, ty

**Comment by pjeweb** [comment]
First time actually using it. Looks great imho.

Screenshot 2024-09-03 at 21 43 25

**Pull request merged**